### PR TITLE
[ASR] Use a subset of manifest when using `scatter` shard strategy

### PR DESCRIPTION
**Pull request by anteju**
# What does this PR do ?

When using `shard_strategy='scatter'` this PR loads only a subset of lines from the manifest file.
This may reduce the time to process a manifest file by an order of magnitude.

Opening as a draft to get feedback if there are any underlying assumptions which may be broken by this change.

**Collection**: ASR

# Changelog 
| File                      | Change |
| ---------------- | -------- |
| `manifest.py::item_iter` | Load only a subset of lines if `shard_strategy == 'scatter'` |
| `collections.py:: ASRAudioText ` | forward `shard_strategy`, `global_rank` and `world_size` to `manifest.item_iter`  |
| `collections.py:: AudioText ` | Use rank and world size to restore the original data list length for `shard_strategy='scatter'` |
|  `audio_to_text.py` | forward `shard_strategy`, `global_rank` and `world_size` to `collections.ASRAudioText` |
| `utils,py` | Added a function to get number of lines from a text file |
| `test_utils.py` | Added unit test for the utility function added above |

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/common/parts/preprocessing/collections.py [hunk 1]
```diff
@@ -190,6 +194,12 @@ def __init__(
             else:
                 data.sort(key=lambda entity: entity.duration)
 
+        if shard_strategy == 'scatter' and world_size > 1:
```
Comment (titu1994) [inline nemo/collections/common/parts/preprocessing/collections.py:197]: Why duplicate data ? It will take multiple times the memory.

File: nemo/collections/common/parts/preprocessing/collections.py [hunk 2]
```diff
@@ -228,7 +248,19 @@ def __init__(self, manifests_files: Union[str, List[str]], *args, **kwargs):
...
+            audio_files,
+            durations,
+            texts,
+            offsets,
+            speakers,
+            orig_srs,
+            token_labels,
+            langs,
+            shard_strategy=shard_strategy,
+            world_size=world_size,
```
Comment (titu1994) [inline nemo/collections/common/parts/preprocessing/collections.py:261]: Missing global rank

File: nemo/collections/common/parts/preprocessing/manifest.py [hunk 3]
```diff
@@ -75,9 +84,44 @@ def item_iter(
...
         cached_manifest_file = DataStoreObject(manifest_file).get()
         logging.debug('Cached at: %s', str(cached_manifest_file))
+
+        if shard_strategy is not None and world_size > 1:
+            if shard_strategy == 'scatter':
+                # Read only necessary manifest lines
+                logging.debug('Partitioning manifest: global rank = %d, world size = %d', global_rank, world_size)
+
+                # Get number of lines in the current manifest
+                num_lines = get_num_lines(cached_manifest_file)
```
Comment (titu1994) [inline nemo/collections/common/parts/preprocessing/manifest.py:94]: Rather import data_util and then explicitly call data_util.get_num_lines for clarity

File: nemo/collections/common/parts/utils.py [hunk 4]
```diff
@@ -96,3 +96,28 @@ def flatten(list_in: List) -> List:
         A flat list of values.
     """
     return list(flatten_iterable(list_in))
+
+
+def get_num_lines(filepath: str, buffer_size: int = 2 ** 16) -> int:
```
Comment (titu1994) [inline nemo/collections/common/parts/utils.py:101]: Name it get num lines from file

**Review by titu1994 — CHANGES_REQUESTED** [review]
Since many classes depend on ASR manifest processor please go through where it is used and update them all. 

Also rework logic to avoid duplicating data

File: nemo/collections/common/parts/preprocessing/collections.py (hunk 1 above)
Comment (titu1994) [reply nemo/collections/common/parts/preprocessing/collections.py:197]: Change the default behavior is better. Such duplication of data just for length calculation is too wasteful

File: tests/collections/common/test_utils.py [hunk 5]
```diff
@@ -31,3 +34,22 @@ def test_flatten(self):
...
+    def test_get_num_lines(self):
+        """Test getting the number of lines from a text file.
+        """
+        test_cases = [0, 1, 10, 42, 97]
+
+        for num_lines in test_cases:
+
+            lines = [f'line {n}\n' for n in range(num_lines)]
+
+            with tempfile.TemporaryDirectory() as tmp_dir:
```
Comment (Kipok) [inline tests/collections/common/test_utils.py:50]: Pytest can handle this natively: https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html (so just add tmp_path fixture and use it insteam of tmp_dir). It's a bit more user-friendly, because if test fails, it's not deleted right away, and you can inspect generated files.

File: tests/collections/common/test_utils.py [hunk 6]
```diff
@@ -31,3 +34,22 @@ def test_flatten(self):
...
         for n, test_case in enumerate(test_cases):
             assert flatten(test_case['input']) == test_case['golden'], f'Test case {n} failed!'
+
+
+class TestTextUtils:
+    @pytest.mark.unit
+    def test_get_num_lines(self):
+        """Test getting the number of lines from a text file.
+        """
+        test_cases = [0, 1, 10, 42, 97]
```
Comment (Kipok) [inline tests/collections/common/test_utils.py:44]: Not very important, but why not put as `@pytest.mark.parametrize(num_lines)`?

File: nemo/collections/common/parts/preprocessing/manifest.py [hunk 7]
```diff
@@ -75,9 +84,44 @@ def item_iter(
...
+            for n, line in enumerate(f):
                 k += 1
+                if begin_idx is not None and n < begin_idx:
+                    # Skip lines before begin_idx
+                    continue
+                if end_idx is not None and n >= end_idx:
+                    # Stop parsing lines from end_idx
+                    break
                 item = parse_func(line, manifest_file)
                 item['id'] = k
```
Comment (Kipok) [inline nemo/collections/common/parts/preprocessing/manifest.py:126]: I'm not really sure what this "id" is used for, but if it's important to preserve the original mapping, then you'd need to continue the loop until the end to make sure it's properly counted, I guess? And if it's not important to have it the same, then you can move the `k += 1` to after the ifs, so that it's going from 0 to N on all processes without skips.

File: nemo/collections/common/parts/preprocessing/manifest.py [hunk 8]
```diff
@@ -75,9 +84,44 @@ def item_iter(
...
+                # Read only necessary manifest lines
+                logging.debug('Partitioning manifest: global rank = %d, world size = %d', global_rank, world_size)
+
+                # Get number of lines in the current manifest
+                num_lines = get_num_lines(cached_manifest_file)
+                logging.debug('Found %d lines in manifest %s', num_lines, cached_manifest_file)
+
+                # Determine start and end lines in the manifest.
+                # This corresponds to splitting the manifest in audio_to_text.py::expand_audio_filepaths
+                begin_idx = (num_lines // world_size) * global_rank
```
Comment (Kipok) [inline nemo/collections/common/parts/preprocessing/manifest.py:99]: I don't fully understand how this works. Are we trying to select all the lines that have shards selected for the current rank? If so, is the assumption here that they always go in sequence and each shard has the same number of lines? I'm just not sure if this is always the case (or if we should assume that it's going to always be the case in the future).

Comment (Kipok) [reply nemo/collections/common/parts/preprocessing/manifest.py:99]: And in any case, I think it's very important to have some tests for this functionality, as it might be very easy to add hard-to-notice bugs here by ignoring part of the data

File: nemo/collections/asr/data/audio_to_text.py [hunk 9]
```diff
@@ -133,6 +133,9 @@ def __init__(
         eos_id: Optional[int] = None,
         pad_id: int = 0,
         index_by_file_id: bool = False,
+        shard_strategy: Optional[str] = None,
+        global_rank: Optional[int] = None,
```
Comment (Kipok) [inline nemo/collections/asr/data/audio_to_text.py:137]: I guess that's more of a question to @titu1994, but is it possible to use pytorch's `dist.get_rank()` and same for world size? Or a light-weight wrapper function if we need to support some edge cases (and in fact I think we already have such function `utils.get_rank`)? Just feels that this is a global property that never changes and would be much easier to get it from a function whenever it's needed, than pass around through each model and dataloader.

File: tests/collections/common/test_utils.py (hunk 5 above)
Comment (titu1994) [reply tests/collections/common/test_utils.py:50]: Did not know this, very nice given how many places we use tempfile in tests

File: nemo/collections/asr/data/audio_to_text.py (hunk 9 above)
Comment (titu1994) [reply nemo/collections/asr/data/audio_to_text.py:137]: I think AppState is a more valid source of "world size" cause it takes a few more cases into account for model parallel. Ignoring that, you first have to check whether distributed is initialized or not, then you can use a method to get world size and local rank ID. It is definitely possible. 

The problem with the global property is the fact that it's global. Originally tarred datasets used such global variables but some domains had case where training is done in multi gpu, you destroy the distributed context and clean up, then eval on single GPU, all in the same run. This caused clash with such cases cause global setting was still set and not cleaned up automatically.

The other way is not via property but global check via testing if pt is under distributed enabled, whether global rank is correct, etc and then return the rank other wise -1. It's higher cost per call though, but it's fine for a init level thing

**Comment by titu1994** [comment]
If it can work for speech ASR dataset (tarred, bucketed) I think @AlexGrinch would like this. Right now we're loss hacking to skip samples
